File: web/src/api/hardware.ts
import { api } from './client'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface DeviceHardware {
  device_id: string
  hostname?: string
  fqdn?: string
  os_name?: string
  os_version?: string
  os_arch?: string
  kernel?: string
  cpu_model?: string
  cpu_cores?: number
  cpu_threads?: number
  cpu_arch?: string
  ram_total_mb?: number
  ram_type?: string
  ram_slots_used?: number
  ram_slots_total?: number
  platform_type?: string
  hypervisor?: string
  vm_host_id?: string
  system_manufacturer?: string
  system_model?: string
  serial_number?: string
  bios_version?: string
  collection_source?: string
  collected_at?: string
  updated_at?: string
}

export interface DeviceStorage {
  id: string
  device_id: string
  name?: string
  disk_type?: string
  interface?: string
  capacity_gb?: number
  model?: string
  role?: string
  collection_source?: string
  collected_at?: string
}

export interface DeviceGPU {
  id: string
  device_id: string
  model?: string
  vendor?: string
  vram_mb?: number
  driver_version?: string
  collection_source?: string
  collected_at?: string
}

export interface DeviceService {
  id: string
  device_id: string
  name?: string
  service_type?: string
  port?: number
  url?: string
  version?: string
  status?: string
  collection_source?: string
  collected_at?: string
}

export interface DeviceHardwareResponse {
  hardware: DeviceHardware | null
  storage: DeviceStorage[]
  gpus: DeviceGPU[]
  services: DeviceService[]
}

export interface HardwareSummary {
  total_with_hardware: number
  total_ram_mb: number
  total_storage_gb: number
  total_gpus: number
  by_os: Record<string, number>
  by_cpu_model: Record<string, number>
  by_platform_type: Record<string, number>
  by_gpu_vendor: Record<string, number>
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

/**
 * Fetch the full hardware profile for a device (hardware, storage, GPUs, services).
 */
export async function getDeviceHardware(id: string): Promise<DeviceHardwareResponse> {
  return api.get<DeviceHardwareResponse>(`/recon/devices/${id}/hardware`)
}

/**
 * Update (or create) a device's hardware profile manually.
 */
export async function updateDeviceHardware(
  id: string,
  data: Partial<DeviceHardware>,
): Promise<DeviceHardware> {
  return api.put<DeviceHardware>(`/recon/devices/${id}/hardware`, data)
}

/**
 * Fetch fleet-wide aggregate hardware statistics.
 */
export async function getHardwareSummary(): Promise<HardwareSummary> {
  return api.get<HardwareSummary>('/recon/inventory/hardware-summary')
}

File: pkg/models/hardware.go
package models

import "time"

// DeviceHardware represents the hardware profile of a device.
type DeviceHardware struct {
	DeviceID           string     `json:"device_id" example:"550e8400-e29b-41d4-a716-446655440000"`
	Hostname           string     `json:"hostname,omitempty" example:"web-server-01"`
	FQDN               string     `json:"fqdn,omitempty" example:"web-server-01.local"`
	OSName             string     `json:"os_name,omitempty" example:"Ubuntu 24.04"`
	OSVersion          string     `json:"os_version,omitempty" example:"24.04"`
	OSArch             string     `json:"os_arch,omitempty" example:"amd64"`
	Kernel             string     `json:"kernel,omitempty" example:"6.5.0-44-generic"`
	CPUModel           string     `json:"cpu_model,omitempty" example:"Intel Core i9-10900K"`
	CPUCores           int        `json:"cpu_cores,omitempty" example:"10"`
	CPUThreads         int        `json:"cpu_threads,omitempty" example:"20"`
	CPUArch            string     `json:"cpu_arch,omitempty" example:"x86_64"`
	RAMTotalMB         int        `json:"ram_total_mb,omitempty" example:"32768"`
	RAMType            string     `json:"ram_type,omitempty" example:"DDR4"`
	RAMSlotsUsed       int        `json:"ram_slots_used,omitempty" example:"2"`
	RAMSlotsTotal      int        `json:"ram_slots_total,omitempty" example:"4"`
	PlatformType       string     `json:"platform_type,omitempty" example:"baremetal"`
	Hypervisor         string     `json:"hypervisor,omitempty" example:"proxmox"`
	VMHostID           string     `json:"vm_host_id,omitempty"`
	SystemManufacturer string     `json:"system_manufacturer,omitempty" example:"Dell Inc."`
	SystemModel        string     `json:"system_model,omitempty" example:"PowerEdge R730"`
	SerialNumber       string     `json:"serial_number,omitempty" example:"ABC123"`
	BIOSVersion        string     `json:"bios_version,omitempty" example:"2.17.0"`
	CollectionSource   string     `json:"collection_source,omitempty" example:"scout-wmi"`
	CollectedAt        *time.Time `json:"collected_at,omitempty"`
	UpdatedAt          *time.Time `json:"updated_at,omitempty"`
}

// DeviceStorage represents a storage device attached to a device.
type DeviceStorage struct {
	ID               string     `json:"id"`
	DeviceID         string     `json:"device_id"`
	Name             string     `json:"name,omitempty" example:"Samsung 990 Pro 4TB"`
	DiskType         string     `json:"disk_type,omitempty" example:"nvme"`
	Interface        string     `json:"interface,omitempty" example:"pcie4"`
	CapacityGB       int        `json:"capacity_gb,omitempty" example:"4000"`
	Model            string     `json:"model,omitempty" example:"Samsung SSD 990 PRO"`
	Role             string     `json:"role,omitempty" example:"data"`
	CollectionSource string     `json:"collection_source,omitempty" example:"scout-linux"`
	CollectedAt      *time.Time `json:"collected_at,omitempty"`
}

// DeviceGPU represents a GPU in a device.
type DeviceGPU struct {
	ID               string     `json:"id"`
	DeviceID         string     `json:"device_id"`
	Model            string     `json:"model,omitempty" example:"NVIDIA RTX 3090 Ti"`
	Vendor           string     `json:"vendor,omitempty" example:"nvidia"`
	VRAMMB           int        `json:"vram_mb,omitempty" example:"24576"`
	DriverVersion    string     `json:"driver_version,omitempty" example:"535.183.01"`
	CollectionSource string     `json:"collection_source,omitempty" example:"scout-linux"`
	CollectedAt      *time.Time `json:"collected_at,omitempty"`
}

// DeviceService represents a running service on a device.
type DeviceService struct {
	ID               string     `json:"id"`
	DeviceID         string     `json:"device_id"`
	Name             string     `json:"name,omitempty" example:"plex"`
	ServiceType      string     `json:"service_type,omitempty" example:"docker"`
	Port             int        `json:"port,omitempty" example:"32400"`
	URL              string     `json:"url,omitempty" example:"http://192.168.1.10:32400"`
	Version          string     `json:"version,omitempty" example:"1.40.0"`
	Status           string     `json:"status,omitempty" example:"running"`
	CollectionSource string     `json:"collection_source,omitempty" example:"scout-linux"`
	CollectedAt      *time.Time `json:"collected_at,omitempty"`
}

// HardwareSummary provides fleet-wide aggregate hardware statistics.
type HardwareSummary struct {
	TotalWithHardware int            `json:"total_with_hardware"`
	TotalRAMMB        int64          `json:"total_ram_mb"`
	TotalStorageGB    int64          `json:"total_storage_gb"`
	TotalGPUs         int            `json:"total_gpus"`
	ByOS              map[string]int `json:"by_os"`
	ByCPUModel        map[string]int `json:"by_cpu_model"`
	ByPlatformType    map[string]int `json:"by_platform_type"`
	ByGPUVendor       map[string]int `json:"by_gpu_vendor"`
}

// HardwareQuery defines filters for hardware-based device queries.
type HardwareQuery struct {
	MinRAMMB     int    `json:"min_ram_mb,omitempty"`
	MaxRAMMB     int    `json:"max_ram_mb,omitempty"`
	CPUModel     string `json:"cpu_model,omitempty"`
	OSName       string `json:"os_name,omitempty"`
	PlatformType string `json:"platform_type,omitempty"`
	GPUVendor    string `json:"gpu_vendor,omitempty"`
	HasGPU       *bool  `json:"has_gpu,omitempty"`
	Limit        int    `json:"limit,omitempty"`
	Offset       int    `json:"offset,omitempty"`
}
